hibike/virtual_device.py

```diff
@@ -0,0 +1,113 @@
+            data = [data_tuple for data_tuple in params_and_values if data_tuple[0] in subscribed_params]
+            hm.send(conn, hm.make_device_data(device_id, data))
+            updateTime = time.time()
+            print("Regular data update sent from %s" % device)
+             
+    msg = hm.read(conn)
+    if not msg:
+        time.sleep(.005)
+        continue
+    if msg.getmessageID() in [hm.messageTypes["SubscriptionRequest"]]:
+#Update the delay, subscription time, and params, then send a subscription response
+        print("Subscription request recieved")
+        params, delay = struct.unpack("<HH", msg.getPayload())
+             
+        subscribed_params = hm.decode_params(device_id, params)
+        hm.send(conn, hm.make_subscription_response(device_id, subscribed_params, delay, uid))
+        updateTime = time.time()
+    if msg.getmessageID() in [hm.messageTypes["Ping"]]: # Send a subscription response 
+             print ("Ping recieved")
+             hm.send(conn, hm.make_subscription_response(device_id, subscribed_params, delay, uid))
+    if msg.getmessageID() in [hm.messageTypes["DeviceRead"]]:
+# Send a device data with the requested param and value tuples
+        print("Device read recieved")
+        params = struct.unpack("<H", msg.getPayload())
+        read_params = hm.decode_params(device_id, params)
+        read_data = 0
+          
+        for data_tuple in params_and_values:
+            if data_tuple[0] in read_params:
+                if not hm.paramMap[device_id][data_tuple[0]][2]:# Raise a syntax error if one of the values to be read is not readable
+                        raise SyntaxError("Attempted to read an unreadable value")
+                read_data.append(data_tuple)
+        hm.send(conn, hm.make_device_data(device_id, read_data))
+           
+    if msg.getmessageID() in [hm.messageTypes["DeviceWrite"]]:
+# Write to requested parameters and return the values of the parameters written to using a device data
+        print("Device write recieved")
+        write_params_and_values = hm.decode_device_write(msg, device_id)
+        write_params = [param_val[0] for param_val in write_params_and_values]
+        value_types = [hm.paramMap[device_id][name][1] for name in write_params]
+                
+        write_tuples = [(write_params[index], write_params_and_values[index][1]) for index in range(len(write_params))]
+        for new_tuple in write_tuples:
+            if not hm.paramMap[device_id][new_tuple[0]][3]: # Raise a syntax error if the value that the message attempted to write to is not writable
+                raise SyntaxError("Attempted to write to an unwritable value")
+            params_and_values[hm.paramMap[device_id][new_tuple[0]][0]] = new_tuple
+
+        # Send the written data, make sure you only send data for readable parameters
+        for index in range(len(write_params)):
+            while not hm.paramMap[device_id][write_tuples[index][0]][2]:
```

Comment:
Does this loop even work correctly? It uses `write_tuples[index][0]` in the while loop condition, but then calls `del write_tuples[index]` in the body. I suspect you will end up entering the while loop once, deleting this object, and then crashing when you evaluate the loop condition a second time. Could you comment on this?